File: epc-domestic-from-la-zips.py
# %%

# use .venv - make sure to restart jupyter kernel. duckdb need python 3.10
import polars as pl
import requests
import yaml
import pyarrow
import duckdb
import get_ca_data as get_ca # functions for retrieving CA \ common data
from pathlib import Path
import os

# %%
def read_api_key():
    with open(os.path.join(os.pardir, 'config.yml'), 'r') as file:
        config = yaml.safe_load(file)
        return config['epc']['auth_token']

# %%

ca_la_df = get_ca.get_ca_la_df(2023, inc_ns=True) # include NS
# ca_la_df.glimpse()

# %%
# retrieve the epc certs from the source csv s
la_list = (ca_la_df['ladcd']) #includes north somerset
f'There are {str(la_list.shape)[1:3]} Local Authorities in Combined Authorities'
# %%
base_zip_url = 'https://epc.opendatacommunities.org/files/domestic-'

# E06000023-Bristol-City-of.zip
#%%

zipfiles_list = [base_zip_url + row['ladcd'] + '-' + row['ladnm'] + '.zip' for row in ca_la_df.iter_rows(named = True)]
zipfiles_list
# %%
def replace_with_hyphen(text):
    return text.replace(' ', '-').replace(',', '')

# %%
zipfiles_urls = [replace_with_hyphen(zipfile) for zipfile in zipfiles_list]

# %%
token = read_api_key()
zipfiles_urls

# %%
token
# %%

# %%
headers = {
    'Accept': 'application/zip',
    'Authorization': f'Basic {token}'
        }

# %%

response = requests.get(zipfiles_urls[0], headers = headers)

response.status_code

# %%
def download_files(url_list, token = token):
 
    headers = {
    'Accept': 'application/zip',
    'Authorization': f'Basic {token}'
        }
    for url in url_list:
        response = requests.get(url, headers = headers)
        filename = 'data/' + url.split("/")[-1]

        with open(filename, 'wb') as file:
            file.write(response.content)

# %%

download_files(zipfiles_urls[0:2])
# %%

File: epc-domestic-from-la-zips.r
pacman::p_load(tidyverse,
               janitor,
               config,
               httr2,
               glue, jsonlite)

token = config::get("auth_token", config = "epc", file = "../config.yml")

headers = list(
  'Accept' =  'application/zip',
  'Authorization' =  glue('Basic {token}')
)


# combined and local authorities

ons_la_ca_url <- "https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/LAD23_CAUTH23_EN_LU/FeatureServer/0/query?where=1%3D1&outFields=*&outSR=4326&f=json"

ca_la_list <- jsonlite::read_json(ons_la_ca_url)

ca_la_tbl <-  ca_la_list %>% 
  pluck("features") %>% 
  enframe() %>% 
  unnest_wider(value) %>% 
  unnest_wider(attributes) %>% 
  rename_with(tolower)


base_zip_url = 'https://epc.opendatacommunities.org/files/domestic-'

# E06000023-Bristol-City-of.zip
#%%

zip_urls <- ca_la_tbl %>% 
  mutate(zip_url = glue("{base_zip_url}{lad23cd}-{lad23nm}.zip") %>% 
           str_remove_all( ",") %>% 
           str_replace_all(" ", "-")) %>% 
  pull(zip_url)

test_bris <- zip_urls[map_lgl(zip_urls, ~grepl("Bristol", .x))]

r <- request(test_bris) %>% 
  req_headers('Accept' =  'application/zip',
              'Authorization' =  glue('Basic {token}')) %>% 
  req_perform()

lobstr::obj_size(r)

resp_status(r)

resp_body_string(r)


download.file(test_bris, destfile = "data/test_bris.zip")
